D8 port.

willtp87 opened the pull request:
Base D8 port.

willtp87 committed:
removing tests and contrib before dmu

willtp87 committed:
update pull template

willtp87 committed:
readme update for d8

willtp87 committed:
dmu run

willtp87 committed:
removing old update hooks

willtp87 committed:
coder and info update

willtp87 committed:
moving forms to d8 structure

willtp87 committed:
syntax error

willtp87 committed:
coder

willtp87 committed:
format string updates

willtp87 committed:
bringing perms around

willtp87 committed:
menu mostly routed

willtp87 committed:
bit of looking at config

willtp87 committed:
less verbose config

willtp87 committed:
form state to d8

willtp87 committed:
db functions are deprecated

willtp87 committed:
smattering of d8 updates

willtp87 committed:
coder

willtp87 committed:
bit of work on config

willtp87 committed:
more config work

willtp87 committed:
coder

willtp87 committed:
admins working

willtp87 committed:
coder

willtp87 committed:
oai endpoint routing

willtp87 committed:
identify works

willtp87 committed:
some d8 db updates

willtp87 committed:
handling config usage updates

adam-vessey commented on config/schema/islandora_oai.formats.schema.yml:
```diff
@@ -0,0 +1,31 @@
+  type: mapping
+  label: Islandora OAI Formats
+  mapping:
+    islandora_oai_metadata_datastream_id:
```
Do... all these really needed to be double-namespace prefixed?

adam-vessey replied:
I... guess it's more convenient for how the form is right now, but... shrug?

adam-vessey replied:
Leaving for larger rework.

adam-vessey commented on islandora_oai.install:
```diff
@@ -5,155 +5,158 @@
+  ];
 
-  $schema['islandora_oai_metadata_formats'] = array(
+  $schema['islandora_oai_metadata_formats'] = [
```
Can this go away now, since we have this in config?

adam-vessey replied:
Leaving for larger rework.

adam-vessey commented on islandora_oai.install:
```diff
@@ -5,155 +5,158 @@
+      'record_prefix' => 'dc',
+      'record_namespace' => 'http://purl.org/dc/elements/1.1/',
+    ])
+    ->execute();
```
All these should be able to go away, yeah? As in, should be in the `config/install` business?

adam-vessey replied:
Leaving for larger rework.

adam-vessey commented on src/Services/Oai.php:
```diff
@@ -0,0 +1,32 @@
+  public function oai() {
+    module_load_include('inc', 'islandora_oai', 'includes/request');
+    $output = islandora_oai_parse_request();
+    $response = new Response();
```
I'm... not quite sure I understand why this was rolled as a service... seems more controller-like, actually generating a response...

... if this is to allow overrides such that other service implementations could be used to override our service implementation, should we then nuke the `hook_islandora_oai_identify_request_handler()` bit? Also, if that's the goal, should probably define the interface explicitly... have a separate controller receive the service via dependency injection, and generate the response there... shrug?

adam-vessey replied:
Addressed in https://github.com/willtp87/islandora_oai/pull/1

adam-vessey commented on src/Form/Admin.php:
```diff
@@ -0,0 +1,225 @@
+    $config->set('islandora_oai_solr_object_ancestors_field', $form_state->getValues()['islandora_oai_configuration']['islandora_oai_solr_object_ancestors_field'])->save();
+    // Because of the dynamic pathing of the OAI path we need to rebuild the
+    // menus.
+    $this->routerBuilder->rebuild();
```
I'm not sure that doing this via the form is sufficient... the config could change in other ways, yeah?

Might be better to have a config change/save listener kind of thing which does the rebuild when our config object changes?

... also... does this do the rebuild synchronously? ... it's before the `->save()`, so it might not see the change in config...

adam-vessey replied:
It _does_ appear to see the change, but... is still spooky?

adam-vessey replied:
Addressed in https://github.com/willtp87/islandora_oai/pull/1

adam-vessey commented on src/Form/Admin.php:
```diff
@@ -0,0 +1,225 @@
+    // Because of the dynamic pathing of the OAI path we need to rebuild the
+    // menus.
+    $this->routerBuilder->rebuild();
+    drupal_set_message($this->t('The configuration options have been saved.'));
```
Somewhat lies, if the `->save()` somehow was to throw an exception, could be confusing seeing a past-tense message.

adam-vessey replied:
Addressed in https://github.com/willtp87/islandora_oai/pull/1

adam-vessey commented on src/Form/HandlerAdmin.php:
```diff
@@ -0,0 +1,344 @@
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    $metadata_formats = [];
+    $results = db_query('SELECT metadata_prefix FROM {islandora_oai_metadata_formats}');
```
Not... quite sure about this business... should be using our config, no?

adam-vessey replied:
Leaving for larger rework.

adam-vessey commented on src/Form/HandlerAdmin.php:
```diff
@@ -0,0 +1,344 @@
+    }
+    // Loop through object linking.
+    $metadata_formats = [];
+    $results = db_query('SELECT metadata_prefix FROM {islandora_oai_metadata_formats}');
```
Table should die?

adam-vessey replied:
Leaving for larger rework.

adam-vessey commented on src/Form/FileManagement.php:
```diff
@@ -0,0 +1,131 @@
+    $form_state->loadInclude('islandora_oai', 'inc', 'includes/handler.admin');
+    // Grab all the user uploaded files.
+    $oai_uploaded_files = [];
+    $upload_path = 'public://islandora_oai_xsls';
```
Not sure where to comment on this, but... should we instead look to move to config for these? Config entities probably, based on `media_entity`?

... Just... XSLTs as (only) managed files couldn't really be exported and imported, methinks... I... _think_ that the `media_entity` stuff could help wrap things up, but... would require some research and experimentation...

adam-vessey replied:
Leaving for larger rework.

adam-vessey commented on config/install/islandora_oai.formats.yml:
```diff
@@ -0,0 +1,33 @@
+formats:
+  mods:
+    islandora_oai_metadata_datastream_id: DC
```
~~Not sure about this... serving DC as MODS?~~

~~... Looks like it [would've _tried_ to be](https://github.com/discoverygarden/islandora_oai/pull/1/files#diff-92ba0f4cf55c781c58b8c3923ae9c2e7L202)... at least [`mods`](https://github.com/discoverygarden/islandora_oai/pull/1/files#diff-6893d226ae73131a45e86cc73a4f376aR42)?... This stuff is spooky...~~ Scratch that... it's reading the variable, which would never have been set...

Has the same issue as the old code, but... do we really want to port the issue? Should we just put `MODS` here?

adam-vessey replied:
Leaving for larger rework.

adam-vessey reviewed (commented):
Oddities using a service... ordering of a few things could be fixed, but... shrug?

adam-vessey committed:
Reorder some method calls...

... be sure to save the config before saying we've done so...
... and to rebuild the router only afterwards...

adam-vessey committed:
Move router rebuilding to an event subscriber.

adam-vessey committed:
Move things over to a controller.

adam-vessey committed:
Avoid caching anything from the dynamic route.

adam-vessey committed:
Document things, and avoid caching.

adam-vessey committed:
Avoid injecting the unused router builder into the admin form.

willtp87 committed:
Merge pull request #1 from adam-vessey/cr-alterations

Code review updates.
